working/1847 - FinalizationRegistry/proposal.md

```diff
@@ -0,0 +1,379 @@
+/// at least one object registered in the registry with that finalization token
+/// is no longer accessible to the program.
+///
+/// If the same object is registered in multiple finalization registries,
+/// or registered multiple times in a single registry,
+/// and the object becomes inaccessible to the program,
+/// then any number of those registrations may trigger their associated
+/// callback. It will not necessarily be all or none of them.
+///
+/// Finalization callbacks will happen as *events*, not during execution of
+/// other code and not as a microtask, but as high-level events similar to
+/// timer events.
+abstract class FinalizationRegistry<F extends Object, FT> {
+  /// Creates a finalization registry with the given finalization callback.
+  external factory FinalizationRegistry(
+    void Function(FT finalizationToken) callback);
+
+  /// Registers [value] for a finalization callback.
+  ///
+  /// When [value] is no longer accessible to the program,
+  /// the registry *may* call its callback function with [finalizationToken]
+  /// as argument.
+  ///
+  /// The [value] and [unregisterToken] arguments do not count towards those
+  /// objects being accessible to the program. Both must be non-[num],
+  /// non-[String], non-[bool], non-[Pointer], non-[Struct] and
+  /// non-[Null] values.
+  ///
+  /// Multiple objects may be registered with the same finalization token,
+  /// and the same object may be registered multiple times with different,
+  /// or the same, finalization token.
+  ///
+  /// The callback may be called at most once per registration, and not
+  /// for registrations which have been unregistered since they were registered.
+  void register(F value, FT finalizationToken, {Object? unregisterToken});
+
+  /// Unregisters any finalization callbacks registered with [unregisterToken]
+  /// as unregister-token.
+  ///
+  /// After unregistering, those callbacks will not happen even if the
+  /// registered object becomes inaccessible.
+  void unregister(Object? unregisterToken);
+}
+
+/// Represents a weak reference to the given [target].
+///
+/// Weak references don't prevent garbage collection from reclaiming an object
+/// which is not reachable through any other (non-[WeakRef]) objects. If
+/// runtime system reclaims such _weakly reachable_ object then all [WeakRef]
+/// objects which reference it will be cleared to contain [null] as their
+/// target.
+abstract class WeakRef {
+  /// Create a [WeakRef] pointing to the given [target].
+  external factory WeakRef(Object? target);
+
+  /// Get the current object pointed by the [WeakRef].
+  external Object? get target;
+
+  /// Set the current object pointed by the [WeakRef].
+  external void set target (Object? newTarget);
+}
+```
+
+The following classes are added to `dart:ffi` library:
+
+```dart
+/// Any variable which has a static type that is a subtype of a [Finalizable]
+/// is guaranteed to be alive for the full duration of a scope in which it is
```

Comment:
Well, it is syntactical duration - I don't see how it can be misinterpreted. (if the execution of the scope did not complete either normally or erroneously then the scope is not complete - this covers both synchronous and asynchronous variants).

I am open for suggestions on how to phrase it.

Reply:
Just saying "syntactic scope" would probably be sufficient, maybe elaborate with "until execution exits the code block where the variable would be in scope". (Function parameters use the body of the function, `try`/`catch` capture variables uses the `catch` block, can't think of other special ways to declare variables. If we introduce if-variables or local-declarations, we'll have to discuss what that means, but they will definitely have a *scope* too where they can be rereferred to.)